keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt

```diff
@@ -944,37 +947,58 @@ class SqlDeliveryConfigRepository(
     val environmentUID = environmentUidByName(deliveryConfigName, environmentName)
       ?: return emptyList()
 
     return sqlRetry.withRetry(READ) {
       jooq
-        .select(ENVIRONMENT_ARTIFACT_CONSTRAINT.ARTIFACT_VERSION)
-        .from(ENVIRONMENT_ARTIFACT_CONSTRAINT)
+        .select(
+          ARTIFACT_VERSIONS.NAME,
+          ARTIFACT_VERSIONS.TYPE,
+          ARTIFACT_VERSIONS.VERSION,
+          ARTIFACT_VERSIONS.RELEASE_STATUS,
+          ARTIFACT_VERSIONS.CREATED_AT,
+          ARTIFACT_VERSIONS.GIT_METADATA,
+          ARTIFACT_VERSIONS.BUILD_METADATA
+        )
+        .from(ENVIRONMENT_ARTIFACT_CONSTRAINT, ARTIFACT_VERSIONS)
         .where(
           ENVIRONMENT_ARTIFACT_CONSTRAINT.ENVIRONMENT_UID.eq(environmentUID),
           ENVIRONMENT_ARTIFACT_CONSTRAINT.STATUS.eq(ConstraintStatus.PENDING.toString())
         )
-        .fetch(ENVIRONMENT_ARTIFACT_CONSTRAINT.ARTIFACT_VERSION)
+        .and(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
+        .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
+        .and(ARTIFACT_VERSIONS.VERSION.eq(ENVIRONMENT_ARTIFACT_CONSTRAINT.ARTIFACT_VERSION))
+        .fetchSortedArtifactVersions(artifact)
     }
   }
 
-  override fun getQueuedConstraintApprovals(deliveryConfigName: String, environmentName: String, artifactReference: String?): Set<String> {
+  override fun getArtifactVersionsQueuedForApproval(deliveryConfigName: String, environmentName: String, artifact: DeliveryArtifact): List<PublishedArtifact> {
     val environmentUID = environmentUidByName(deliveryConfigName, environmentName)
-      ?: return emptySet()
+      ?: return emptyList()
 
     return sqlRetry.withRetry(READ) {
-      jooq.select(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_VERSION)
-        .from(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL)
+      jooq.select(
+        ARTIFACT_VERSIONS.NAME,
+        ARTIFACT_VERSIONS.TYPE,
+        ARTIFACT_VERSIONS.VERSION,
+        ARTIFACT_VERSIONS.RELEASE_STATUS,
+        ARTIFACT_VERSIONS.CREATED_AT,
+        ARTIFACT_VERSIONS.GIT_METADATA,
+        ARTIFACT_VERSIONS.BUILD_METADATA
+      )
+        .from(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL, ARTIFACT_VERSIONS)
         .where(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ENVIRONMENT_UID.eq(environmentUID))
         .and(
-          ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_REFERENCE.eq(artifactReference)
+          ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_REFERENCE.eq(artifact.reference)
             // for backward comparability
             .or(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_REFERENCE.isNull)
         )
-        .fetch(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_VERSION)
-        .toSet()
+        .and(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
+        .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
+        .and(ARTIFACT_VERSIONS.VERSION.eq(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_VERSION))
+        .fetchSortedArtifactVersions(artifact)
```

Comment:
This applies native SQL sorting and limiting when applicable, and falls back to the comparator from the sorting strategy when that doesn't apply. This means the return from this function is pre-sorted.